Lab1_Operationalizing_Pytorch_with_Mlflow/src/bert-classifier/news_classifier.py

```diff
@@ -0,0 +1,412 @@
+                attention_mask = d["attention_mask"].to(self.device)
+                targets = d["targets"].to(self.device)
+
+                outputs = model(input_ids=input_ids, attention_mask=attention_mask)
+                _, preds = torch.max(outputs, dim=1)
+
+                probs = F.softmax(outputs, dim=1)
+
+                review_texts.extend(texts)
+                predictions.extend(preds)
+                prediction_probs.extend(probs)
+                real_values.extend(targets)
+
+        predictions = torch.stack(predictions).cpu()
+        prediction_probs = torch.stack(prediction_probs).cpu()
+        real_values = torch.stack(real_values).cpu()
+        return review_texts, predictions, prediction_probs, real_values
+
+
+if __name__ == "__main__":
+
+    parser = argparse.ArgumentParser(description="PyTorch BERT Example")
+
+    parser.add_argument(
+        "--max_epochs",
+        type=int,
+        default=5,
+        metavar="N",
+        help="number of epochs to train (default: 14)",
+    )
+
+    parser.add_argument(
+        "--num_samples",
+        type=int,
+        default=15000,
+        metavar="N",
+        help="Number of samples to be used for training "
+        "and evaluation steps (default: 15000) Maximum:100000",
+    )
+
+    parser.add_argument(
+        "--save_model",
+        type=bool,
+        default=True,
+        help="For Saving the current Model",
+    )
+
+    parser.add_argument(
+        "--vocab_file",
+        default="https://s3.amazonaws.com/models.huggingface.co/bert/bert-base-uncased-vocab.txt",
+        help="Custom vocab file",
+    )
+
+    parser.add_argument(
+        "--model_save_path", type=str, default="models", help="Path to save mlflow model"
+    )
+
+    args = parser.parse_args()
+    mlflow.start_run()
+
+    model = NewsClassifier(args)
+    model = model.to(model.device)
+    model.prepare_data()
+    model.setOptimizer()
+    model.startTraining(model)
+
+    print("TRAINING COMPLETED!!!")
+
+    test_acc, test_loss = model.eval_model(model, model.test_data_loader)
+
+    print(test_acc.item())
+
+    y_review_texts, y_pred, y_pred_probs, y_test = model.get_predictions(
+        model, model.test_data_loader
+    )
+
+    # Logging and registering
+    mlflow.log_param("epochs", model.EPOCHS)
+    mlflow.log_param("samples", model.NUM_SAMPLES_COUNT)
+    mlflow.log_metric("test_acc", float(test_acc))
+    mlflow.log_metric("test_loss", float(test_loss))
+    mlflow.pytorch.log_model(model, "bert-model", registered_model_name="BertModel")
+
+    print("\n\n\n SAVING MODEL")
+
+    if args.save_model:
+        if os.path.exists(args.model_save_path):
+            shutil.rmtree(args.model_save_path)
+        mlflow.pytorch.save_model(
+            model,
+            path=args.model_save_path,
+            requirements_file="requirements.txt",
+            extra_files=["class_mapping.json", "bert_base_uncased_vocab.txt"],
+        )
+    
```

Comment:
```suggestion
        mlflow.log_artifacts(args.model_save_path, "state_dict_version")
```
If you want to allow saving the state dict version of the model so that it will work with the cloud-based mlflow server